### PR TITLE
feat: implement new download URL variable `${code}`

### PR DESCRIPTION
**Pull request by smira**
New variable value is coming from `META`, and it might be set using the interactive console (not implemented yet, but it will come soon).

I had to refactor the URL expansion implementation:

* simplify things where possible
* provide more unit-tests for smaller units
* handle expansion of all variables in parallel
* allow parallel expansion on multiple variables

Also I refactored download code to support proper passing of endpoint
function with context.

The end result:

* Talos will try to download config for 3 hours before rebooting
* Each attempt which includes URL expansion + download is limited to 3
  minutes

### CODE REVIEW

File: internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"fmt"
+	"net/url"
+
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/slices"
+)
+
+// Populate populates the config download URL with values replacing variables.
+func Populate(ctx context.Context, downloadURL string, st state.State) (string, error) {
+	return PopulateVariables(ctx, downloadURL, st, AllVariables())
```
Comment (frezbo) [inline internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url.go:19]: same function signatures? and I don;t see `PopulateVariables` being used as is :thinking: ?

**Review by frezbo — APPROVED** [review]

File: internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url.go (hunk 1 above)
Comment (smira) [reply internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url.go:19]: I kind of did more generic version which accepts custom variable list...

**Comment by smira** [comment]
Sample log:

```
[    6.638535] [talos] phase config (10/11): 1 tasks(s)
[    6.639149] [talos] task loadConfig (1/1): starting
[    6.639708] [talos] task loadConfig (1/1): downloading config
[    6.640349] [talos] waiting for network to be ready
[    6.641129] [talos] waiting for URL variables: [code]
[    6.641739] [talos] waiting for URL variables: [code]
[    6.661455] [talos] adjusting time (slew) by 80.803927ms via 188.93.90.21, state TIME_OK, status STA_PLL | STA_NANO {"component": "controller-runtime", "controller": "time.SyncController"}
[   96.561979] [talos] failed to populate talos.config fetch URL "http://172.20.0.1:37159/config.yaml?code=${code}": context deadline exceeded
[   96.563384] [talos] fetching machine config from: "http://172.20.0.1:37159/config.yaml?code=%24%7Bcode%7D"
[   96.566271] [talos] retrying error: failed to download config from "http://172.20.0.1:37159/config.yaml?code=${code}": failed to download config, received 404
[   97.017000] [talos] waiting for URL variables: [code]
[   97.017661] [talos] waiting for URL variables: [code]
[  187.016226] [talos] failed to populate talos.config fetch URL "http://172.20.0.1:37159/config.yaml?code=${code}": context deadline exceeded
[  187.017641] [talos] fetching machine config from: "http://172.20.0.1:37159/config.yaml?code=%24%7Bcode%7D"
[  188.039588] [talos] waiting for URL variables: [code]
[  188.040186] [talos] waiting for URL variables: [code]
[  278.038566] [talos] failed to populate talos.config fetch URL "http://172.20.0.1:37159/config.yaml?code=${code}": context deadline exceeded
[  278.039976] [talos] fetching machine config from: "http://172.20.0.1:37159/config.yaml?code=%24%7Bcode%7D"
[  281.332169] [talos] waiting for URL variables: [code]
[  281.332807] [talos] waiting for URL variables: [code]
[  371.331245] [talos] failed to populate talos.config fetch URL "http://172.20.0.1:37159/config.yaml?code=${code}": context deadline exceeded
[  371.332610] [talos] fetching machine config from: "http://172.20.0.1:37159/config.yaml?code=%24%7Bcode%7D"
[  378.927656] [talos] waiting for URL variables: [code]
[  378.928298] [talos] waiting for URL variables: [code]
[  468.926600] [talos] failed to populate talos.config fetch URL "http://172.20.0.1:37159/config.yaml?code=${code}": context deadline exceeded
[  468.928045] [talos] fetching machine config from: "http://172.20.0.1:37159/config.yaml?code=%24%7Bcode%7D"
[  484.284222] [talos] waiting for URL variables: [code]
[  484.284866] [talos] waiting for URL variables: [code]
...
```

File: internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url_test.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		},
+		{
+			name:     "multiple variables",
+			url:      "https://example.com?uuid=${uuid}&mac=${mac}&hostname=${hostname}&code=${code}",
+			expected: "https://example.com?code=top-secret&hostname=example-node&mac=12%3A34%3A56%3A78%3A90%3Aab&uuid=0000-0000",
+			preSetup: []setupFunc{
+				createSysInfo("0000-0000", "12345"),
+				createMac("12:34:56:78:90:ab"),
+				createHostname("example-node"),
+				createCode("top-secret"),
```
Comment (andrewrynhard) [inline internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url_test.go:71]: 😄

**Review by andrewrynhard — APPROVED** [review]

**Review by frezbo — APPROVED** [review]

**Review by DmitriyMV — APPROVED** [review]

File: internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+			}()
+
+			for _, f := range test.parallelSetup {
+				f(ctx, t, st)
+			}
+
+			err := <-errCh
+			require.NoError(t, err)
+
+			assert.Equal(t, test.expected, result)
```
[the remark this answers is not in the document]
Comment (smira) [reply internal/app/machined/pkg/runtime/v1alpha1/platform/metal/internal/url/url_test.go:122]: hmm... not sure I got it...

**Pull request merged**